cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -54,76 +166,271 @@ theorem not_contains_of_empty {α β} [BEq α] (k : α) :
 := by simp [make, empty, List.canonicalize_nil]
 
+/--
+  Note that the converse of this is not true:
+  counterexample `xs = [(1, false)]`, `ys = []`, `ab = (1, false)`.
+-/
+theorem make_cons [LT α] [DecidableLT α] {xs ys : List (α × β)} {ab : α × β} :
+  make xs = make ys → make (ab :: xs) = make (ab :: ys)
+:= by
+  simp only [make, mk.injEq]
+  apply List.canonicalize_cons
+
 /-! ### find? and mapOnValues -/
 
-theorem find?_mapOnValues {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) (m : Map α β) (k : α)  :
+theorem mapOnValues_wf [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} :
+  m.WellFormed ↔ (m.mapOnValues f).WellFormed
+:= by
+  unfold mapOnValues WellFormed toList
+  sorry
+
+theorem mapOnValues_empty {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} :
+  (empty : Map α β).mapOnValues f = empty
+:= by
+  simp [mapOnValues, empty]
+
+theorem find?_mapOnValues {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α}  :
   (m.find? k).map f = (m.mapOnValues f).find? k
 := by
-  simp [Map.find?, Map.mapOnValues, Map.kvs, ←List.find?_pair_map]
-  cases __ : List.find? (fun x => x.fst == k) m.1 <;>
-  simp only [Option.map_none', Option.map_some']
+  simp only [find?, kvs, mapOnValues, ← List.find?_pair_map]
+  cases m.1.find? (λ x => x.fst == k) <;> simp
 
-theorem find?_mapOnValues_some {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) {m : Map α β} {k : α} {v : β} :
+theorem find?_mapOnValues_some {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} {v : β} :
   m.find? k = .some v →
   (m.mapOnValues f).find? k = .some (f v)
 := by
   intro h₁
-  rw [← find?_mapOnValues f m k]
+  rw [← find?_mapOnValues]
   simp [Option.map, h₁]
 
-theorem find?_mapOnValues_none {α β γ} [LT α] [DecidableLT α] [DecidableEq α] (f : β → γ) {m : Map α β} {k : α} :
+theorem find?_mapOnValues_none {α β γ} [LT α] [DecidableLT α] [DecidableEq α] {f : β → γ} {m : Map α β} {k : α} :
   m.find? k = .none →
   (m.mapOnValues f).find? k = .none
 := by
   intro h₁
-  rw [← find?_mapOnValues f m k]
+  rw [← find?_mapOnValues]
   simp [Option.map, h₁]
 
-theorem mapOnValues_eq_make_map {α β γ} [LT α] [StrictLT α] [DecidableLT α] (f : β → γ) {m : Map α β}
-  (h₁ : m.WellFormed) :
+theorem mapOnValues_eq_make_map {α β γ} [LT α] [StrictLT α] [DecidableLT α] {f : β → γ} {m : Map α β}
+  (wf : m.WellFormed) :
   m.mapOnValues f = Map.make (m.toList.map λ kv => (kv.fst, f kv.snd))
 := by
-  simp [Map.WellFormed] at h₁
-  simp [Map.mapOnValues, Map.toList, Map.kvs, Map.make] at *
-  rw [h₁] ; simp only ; rw [eq_comm]
-  have h₂ : Prod.map id f = (fun (x : α × β) => (x.fst, f x.snd)) := by unfold Prod.map ; simp only [id_eq]
-  simp only [← h₂, ← List.canonicalize_of_map_fst, List.canonicalize_idempotent]
-
-theorem find?_mem_toList {α β} [LT α] [DecidableLT α] [DecidableEq α] {m : Map α β} {k : α} {v : β}
-  (h₁ : m.find? k = .some v) :
-  (k, v) ∈ m.toList
```

Comment:
I needed the converse of this as well, so I reformulated as an `iff`, renamed to `in_list_some_find?`, and reformulated using `.kvs` instead of `.toList` (which are synonymous, I'm not sure why we have both).  I needed to introduce a well-formedness precondition, but that wasn't a problem at any existing usage sites.

Reply:
It breaks a large internal development.  Let's keep the old definitions.  

As a general rule, we want to keep the shared files as stable as possible, and not make breaking changes unless really necessary.

We can do a cleanup pass once our proofs are stable :)

Reply:
I'll revert.

Reply:
Reverted and revised this PR